File: LeetCodeSolutions/LeetCode_0209.py
class Solution:
    def minSubArrayLen(self, s: int, nums: List[int]) -> int:
        if not nums:
            return 0
        l, r = 0, 0
        ans = nums[0]
        cnt = len(nums)
        isvalid = False
        while l < len(nums) and r < len(nums):
            if ans < s:
                if r == len(nums) - 1:
                    break
                r += 1
                ans += nums[r]
            else:
                isvalid = True
                cnt = min(cnt, r - l + 1)
                ans -= nums[l]
                l += 1
            # print(l, r, ans, cnt)
        return cnt if isvalid else 0

File: LeetCodeSolutions/LeetCode_0209.cpp
class Solution {
public:
    int minSubArrayLen(int s, vector<int>& nums) {
        int l = 0, r = 0, total = 0;
        int ans  = INT_MAX;
        while(r < nums.size()){
            while (r < nums.size() && total < s){
                total += nums[r];
                ++r;
            }
            while(l < r && total - nums[l] >= s){
                total -= nums[l];
                ++l;
            }
            if(total >= s) ans = min(ans, r - l);
            total -= nums[l];
            ++l;
        }
        return ans == INT_MAX? 0: ans;
    }
};